content/2019-02-19_osmnacty-sraz-git.md

```diff
@@ -0,0 +1,71 @@
+
+### git cherry-pick
+
+Jedná se o příkaz, který umožňuje výběr konkrétních commitů libovolně z historie. 
+`git cherry-pick` spolu s kódem commitu použije jen tento vybraný commit a udělá z něj nový commit na aktuální větvi, aplikuje změny z daného commitu. Pokud místo commitu napíšu název větve, je to jen jiný název commitu, na který daná větev ukazuje. Nevezme větev celou. 
+Více k tomuto tématu viz [dokumentace](https://git-scm.com/docs/git-cherry-pick).
+
+### git reset
+
+Příkaz `git reset` umí rušit commity. Pokud za příkaz zadáme název commitu, přesuneme se na něj a tedy v rámci gitu do stavu, v jakém tehdy repozitář byl. Změny, které se staly v pozdějších commitech, nejsou nyní v gitu, ale existují v našem repozitáři. Pokud se jich chci zbavit a úplně se vrátit do předchozího stavu, existují dvě cesty. Buď po `git reset "název commitu"` zadám ještě `git checkout .` Druhá cesta je přímější, a to pomocí `git reset --hard "název commitu"` přímo pozmění repozitář do stavu v daném commitu a vše, co se stalo po něm, zahodí, včetně rozpracovaných věcí. 
+Pomocníkem při pohybech po commitech může být `git reflog`. Git si pamatuje, jak se posouval ukazatel `HEAD` v rámci repozitáře, a `git reflog` vypíše historii všech těchto pohybů, včetně vymazaných commitů a také včetně provedených příkazů, které k pohybu vedly. Historie se ale čas od času promazává, nelze tedy jít libovolně dávno do minulosti.
+
+![git_reflog](./images/reflog.png)
+
+V rámci gitu se pohybjeme v podstatě ve třech úrovních - viz obrázek níže. Vycházíme z našeho pracovního adresáře, kde máme vztvořený gitový repozitář. Dokud gitu neřekneme pomocí `git add`, aby sledoval nějaké změny, červenou barvou nám po zadání `git status` vypíše soubory, které v adresáři jsou, ale nesleduje je. Když vložíme soubor do gitu, přesune se do zelené oblasti, tzv. Stage, kde je již sleduje, ale stále s nimi můžeme pracovat, měnit je apod. 
+Pokud máme soubor ve Stagi, ale chceme jej odebrat a vrátit zpět do pracovního adresáře, lze to pomocí `git reset -- "název souboru"`. 
+Ze Stage pak pomocí příkazu `git commit` git sledované soubory zabalí, dá jim identifikátor, hlavičku, datum, email apod., vytvoří commit a uloží ho. 
+
+![git_colors](./images/git_colors.png)
+*[Zdroj: Začátečnický kurz PyLadies](https://pyvec.github.io/cheatsheets/basic-git/basic-git-cs.pdf)*
+
+### git rebase
+
+Příkaz `git rebase` vykonává opakovaný `git cherry-pick`. Zadáme-li `git rebase "název větve"`, git se podívá, jak se od sebe dané větve liší a jaké jsou mezi nimi rozdílné commity. Postupně pak jednotlivé commity vezme a provede `git cherry-pick`. Důležité je pořadí větví. Ta větev, na které aktuálně jsem, slouží jako zdroj commit. Větev, kterou předáme jako parametr příkazu `git rebase`, bude ta, na kterou budou commity přibývat. 
```

Comment:
> Ta větev, na které aktuálně jsem, slouží jako zdroj commit. Větev, kterou předáme jako parametr příkazu `git rebase`, bude ta, na kterou budou commity přibývat.

Zde ocenim primy priklad: pokud pracuji na vetvi moje-zmeny, a mezitim mi ujel master, mam se prepnout do vetve moje-zmeny a rict: git rebase master, nebo naopak? (Pokazde si to hledam, pokazde se to pokousim pouzit blbe ;)).

Reply:
Koukni na ten link, co jsem dávala na závěr téhle části, která to má i rozkreslené. Proto jsem to tam dávala, abych nemusela přepisovat něco, co už někdo sepsal i s grafickým příkladem ;)

Reply:
U tvého příkladu: pokud chceš přesunout větev moje-zmeny na začátek masteru a aktuálně jsi na větvi moje-zmeny, zůstaneš na ní a dáš  `git rebase master`, alespoň tak jsem to pochopila z dokumentace.
Na lekci to právě bylo vysvětlené hodně zmateně, poslouchala jsem to několikrát a kontrolovala si to v dokumentaci a různých návodech, abych zjistila, že to je naopak, než bylo prvně řečeno na lekci.

Reply:
Ok, mrknu na odkazy. Myslím, že tam to stačí.

Reply:
Ano, když sedím na větvi `moje-zmeny`, tak `git rebase master` přeroubuje aktuální větev (`moje-zmeny`) na `master`. V podstatě to má syntaxi `git rebase <nekam>`. `master` zůstane netknutný.